lib/ret/uploads.ex

```diff
@@ -0,0 +1,140 @@
+defmodule Ret.Uploads do
+  require Logger
+
+  @chunk_size 1024 * 1024
+
+  # Given a Plug.Upload and an optional encryption key, returns an id
+  # that can be used to fetch a stream to the uploaded file after this call.
+  def store(%Plug.Upload{content_type: content_type, filename: filename, path: path}, key) do
+    with uploads_storage_path when is_binary(uploads_storage_path) <- module_config(:storage_path) do
+      {:ok, %{size: content_length}} = File.stat(path)
+      uuid = Ecto.UUID.generate()
+
+      [upload_path, meta_file_path, blob_file_path] = paths_for_uuid(uuid)
+      File.mkdir_p!(upload_path)
+
+      case write_blob_file(path, blob_file_path, key) do
+        :ok ->
+          meta = %{
+            content_type: content_type,
+            content_length: content_length,
+            filename: filename,
+            blob: blob_file_path,
+            encrypted: key != nil
+          }
+
+          meta_file_path |> File.write!(Poison.encode!(meta))
+
+          {:ok, uuid}
+
+        {:error, _reason} = err ->
+          err
+      end
+    else
+      _ -> {:error, :not_allowed}
+    end
+  end
+
+  def fetch(id, key) do
+    with uploads_storage_path when is_binary(uploads_storage_path) <- module_config(:storage_path) do
+      case Ecto.UUID.cast(id) do
+        {:ok, uuid} ->
+          [_upload_path, meta_file_path, blob_file_path] = paths_for_uuid(uuid)
+
+          case [File.stat(meta_file_path), File.stat(blob_file_path)] do
+            [{:ok, _stat}, {:ok, _blob_stat}] ->
+              meta = File.read!(meta_file_path) |> Poison.decode!()
+
+              case read_blob_file(blob_file_path, meta, key) do
+                {:ok, stream} -> {:ok, meta, stream}
+                {:error, :invalid_key} -> {:error, :not_allowed}
+                _ -> {:error, :not_found}
+              end
+
+            _ ->
+              {:error, :not_found}
+          end
+
+        :error ->
+          {:error, :not_found}
+      end
+    else
+      _ -> {:error, :not_allowed}
+    end
+  end
+
+  # Vacuums up TTLed out uploads
+  def vacuum do
+    Logger.info("Uploads: Beginning Vacuum.")
+
+    with uploads_storage_path when is_binary(uploads_storage_path) <-
+           module_config(:storage_path),
+         uploads_ttl when is_integer(uploads_ttl) <- module_config(:ttl) do
+      process_meta = fn meta_file, _acc ->
+        meta = File.read!(meta_file) |> Poison.decode!()
+        blob_file = meta["blob"]
+
+        {:ok, %{atime: atime}} = File.stat(blob_file)
```

Comment:
Are we sure `atime` is going to behave as expected? From what I've read, it depends on how the filesystem was mounted.

Reply:
I will double check -- we're using EFS over NFS so good call making sure it does what we want.